### PR TITLE
DownloadController: Partial Compose conversion

### PR DESCRIPTION
**Pull request by ivaniskandar**
Item list is not changed as currently there is no fitting Compose component to replace the drag-drop behavior.

CI build most likely will fail to complete :/

### CODE REVIEW

**Comment by ivaniskandar** [comment]
Interesting

File: app/src/main/java/eu/kanade/tachiyomi/ui/download/DownloadHeaderItem.kt [hunk 1]
```diff
@@ -35,16 +35,39 @@ data class DownloadHeaderItem(
...
     override fun hashCode(): Int {
-        return id.hashCode()
+        var result = id.hashCode()
+        result = 31 * result + name.hashCode()
+        result = 31 * result + size
+        result = 31 * result + subItems.hashCode()
+        return result
     }
 
+//    override fun equals(other: Any?): Boolean {
```
Comment (arkon) [inline app/src/main/java/eu/kanade/tachiyomi/ui/download/DownloadHeaderItem.kt:59]: Remove these.

File: app/src/main/java/eu/kanade/tachiyomi/data/download/DownloadService.kt [hunk 2]
```diff
@@ -47,6 +49,13 @@ class DownloadService : Service() {
          */
         val runningRelay: BehaviorRelay<Boolean> = BehaviorRelay.create(false)
 
+        private val _isRunning = MutableStateFlow(false)
+
+        /**
+         * True if the service is running.
+         */
```
Comment (arkon) [inline app/src/main/java/eu/kanade/tachiyomi/data/download/DownloadService.kt:56]: Is this really useful...?

File: app/src/main/java/eu/kanade/tachiyomi/ui/download/DownloadController.kt [hunk 3]
```diff
@@ -1,134 +1,318 @@
...
-        DownloadService.runningRelay
-            .observeOn(AndroidSchedulers.mainThread())
-            .subscribeUntilDestroy { onQueueStatusChange(it) }
+    @Composable
+    override fun ComposeContent() {
+        val context = LocalContext.current
+        val downloadList by presenter.state.collectAsState()
+
+        val scrollBehavior = TopAppBarDefaults.pinnedScrollBehavior(rememberTopAppBarState())
+        var fabExpanded by remember { mutableStateOf(true) }
```
Comment (arkon) [inline app/src/main/java/eu/kanade/tachiyomi/ui/download/DownloadController.kt:111]: Really ought to lift this to our `Scaffold` or something at some point.

File: app/src/main/java/eu/kanade/tachiyomi/ui/download/DownloadController.kt [hunk 4]
```diff
@@ -1,134 +1,318 @@
...
+                        Row(verticalAlignment = Alignment.CenterVertically) {
+                            Text(
+                                text = stringResource(R.string.label_download_queue),
+                                maxLines = 1,
+                                modifier = Modifier.weight(1f, false),
+                                overflow = TextOverflow.Ellipsis,
+                            )
+                            if (downloadList.isNotEmpty()) {
+                                val pillAlpha = if (isSystemInDarkTheme()) 0.12f else 0.08f
+                                Pill(
```
Comment (arkon) [inline app/src/main/java/eu/kanade/tachiyomi/ui/download/DownloadController.kt:147]: I feel like we could also just make `AppBar` accept something for a pill considering it's a pattern we have in the library too.

File: app/src/main/java/eu/kanade/tachiyomi/ui/download/DownloadController.kt [hunk 5]
```diff
@@ -1,134 +1,318 @@
...
+        ) { contentPadding ->
+            val density = LocalDensity.current
+            val layoutDirection = LocalLayoutDirection.current
+            val left = with(density) { contentPadding.calculateLeftPadding(layoutDirection).toPx().roundToInt() }
+            val top = with(density) { contentPadding.calculateTopPadding().toPx().roundToInt() }
+            val right = with(density) { contentPadding.calculateRightPadding(layoutDirection).toPx().roundToInt() }
+            val bottom = with(density) { contentPadding.calculateBottomPadding().toPx().roundToInt() }
+
+            Box(modifier = Modifier.nestedScroll(nestedScrollConnection)) {
+                if (downloadList.isEmpty()) {
```
Comment (arkon) [inline app/src/main/java/eu/kanade/tachiyomi/ui/download/DownloadController.kt:275]: Couldn't we do the empty state before the rest of it and short circuit?

File: app/src/main/java/eu/kanade/tachiyomi/data/download/DownloadService.kt (hunk 2 above)
Comment (ivaniskandar) [reply app/src/main/java/eu/kanade/tachiyomi/data/download/DownloadService.kt:56]: As in the value? It's used for start/pause FAB. But yeah with this, there's a three way to check for the service status. So it's kinda all over the place

Comment (arkon) [reply app/src/main/java/eu/kanade/tachiyomi/data/download/DownloadService.kt:56]: I meant the Javadoc

Comment (ivaniskandar) [reply app/src/main/java/eu/kanade/tachiyomi/data/download/DownloadService.kt:56]: Ah. Now that I look at it again no

File: app/src/main/java/eu/kanade/tachiyomi/ui/download/DownloadController.kt (hunk 3 above)
Comment (ivaniskandar) [reply app/src/main/java/eu/kanade/tachiyomi/ui/download/DownloadController.kt:111]: This one's specific for interop nested scroll. A simpler method by using LazyList state is available and IMO should be kept separate.

**Pull request merged**